// @typegoose/auto-increment/lib/autoIncrement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoIncrementID = exports.AutoIncrementSimple = void 0;
const tslib_1 = require("tslib");
const lodash_1 = require("lodash");
const mongoose = require("mongoose");
const util_1 = require("util");
const logSettings_1 = require("./logSettings");
const DEFAULT_INCREMENT = 1;
/**
 * The Plugin - Simple version
 * Increments an value each time it is saved
 * @param schema The Schema
 * @param options The Options
 */
function AutoIncrementSimple(schema, options) {
    // convert normal object into an array
    const fields = Array.isArray(options) ? options : [options];
    logSettings_1.logger.info('Initilaize AutoIncrement for an schema with %d fields to increment', fields.length);
    if (fields.length <= 0) {
        throw new Error('Options with at least one field are required!');
    }
    // check if all fields are valid
    for (const field of fields) {
        const schemaField = schema.path(field.field);
        // check if the field is even existing
        if (util_1.isNullOrUndefined(schemaField)) {
            throw new Error(`Field "${field.field}" does not exists on the Schema!`);
        }
        // check if the field is an number
        if (!(schemaField instanceof mongoose.Schema.Types.Number)) {
            throw new Error(`Field "${field.field}" is not an SchemaNumber!`);
        }
        if (util_1.isNullOrUndefined(field.incrementBy)) {
            logSettings_1.logger.info('Field "%s" does not have an incrementBy defined, defaulting to %d', field.field, DEFAULT_INCREMENT);
            field.incrementBy = DEFAULT_INCREMENT;
        }
    }
    schema.pre('save', function AutoIncrementPreSaveSimple() {
        if (!this.isNew) {
            logSettings_1.logger.info('Starting to increment "%s"', this.constructor.modelName);
            for (const field of fields) {
                logSettings_1.logger.info('Incrementing "%s" by %d', field.field, field.incrementBy);
                this[field.field] += field.incrementBy;
            }
        }
    });
}
exports.AutoIncrementSimple = AutoIncrementSimple;
/** The Schema used for the trackers */
const IDSchema = new mongoose.Schema({
    field: String,
    model: String,
    count: Number
}, { versionKey: false });
IDSchema.index({ field: 1, model: 1 }, { unique: true });
/**
 * The Plugin - ID
 * Increments an counter in an tracking collection
 * @param schema The Schema
 * @param options The Options
 */
function AutoIncrementID(schema, options) {
    /** The Options with default options applied */
    const opt = lodash_1.merge({}, {
        field: '_id',
        incrementBy: DEFAULT_INCREMENT,
        trackerCollection: 'identitycounters',
        trackerModelName: 'identitycounter',
        startAt: 0
    }, options);
    // check if the field is an number
    if (!(schema.path(opt.field) instanceof mongoose.Schema.Types.Number)) {
        throw new Error(`Field "${opt.field}" is not an SchemaNumber!`);
    }
    let model;
    logSettings_1.logger.info('AutoIncrementID called with options %O', opt);
    schema.pre('save', function AutoIncrementPreSaveID() {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            logSettings_1.logger.info('AutoIncrementID PreSave');
            if (!this.isNew) {
                logSettings_1.logger.info('Document is not new, not incrementing');
                return;
            }
            const modelName = this.constructor.modelName;
            if (!model) {
                logSettings_1.logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
                // needs to be done, otherwise "undefiend" error if the plugin is used in an sub-document
                const db = (_a = this.db) !== null && _a !== void 0 ? _a : this.ownerDocument().db;
                model = db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
                // test if the counter document already exists
                const counter = yield model.findOne({ model: modelName, field: opt.field }).lean().exec();
                if (!counter) {
                    yield model.create({
                        model: modelName,
                        field: opt.field,
                        count: opt.startAt - opt.incrementBy
                    });
                }
            }
            const { count } = yield model.findOneAndUpdate({
                field: opt.field,
                model: modelName
            }, {
                $inc: { count: opt.incrementBy }
            }, {
                new: true,
                fields: { count: 1, _id: 0 },
                upsert: true,
                setDefaultsOnInsert: true
            }).lean().exec();
            logSettings_1.logger.info('Setting "%s" to "%d"', opt.field, count);
            this[opt.field] = count;
            return;
        });
    });
}
exports.AutoIncrementID = AutoIncrementID;
tslib_1.__exportStar(require("./types"), exports);
